File: src/AlgorithmAnalysis/DoublingTest.java
package AlgorithmAnalysis;

// Equation of log-log plot:
//  T(n) = an^3, where a = 9.98 * 10^-11

import java.util.Random;

public class DoublingTest {
    public static double TimeTrial(int n) {
        int MAX = 1000000;
        Random random = new Random();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = random.nextInt(MAX + MAX) - MAX;
        }
        long start = System.currentTimeMillis();
        int count = count(arr);
        long end = System.currentTimeMillis();
        return (end - start) / 1000.0;
    }

    public static int count(int[] a) {
        int n = a.length;
        int count = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                for (int k = j+1; k < n; k++) {
                    if (a[i] + a[j] + a[k] == 0) count++;
                }
            }
        }
        return count;
    }

    //Test Client
    public static void main(String[] args) {
        int upperBound = Integer.parseInt(args[0]);
        for (int n = 250; n < upperBound; n *= 2) {
            double time = TimeTrial(n);
            System.out.println("N: " + n + " TIME: " + time);
        }
    }
}

File: src/Kotlin/AlgorithmAnalysis/DoublingTest.kt
package Kotlin.AlgorithmAnalysis

import java.util.Random

// Doubling Test.

class DoublingTest {

    fun timeTrial(n: Int): Double {
        val start: Long = System.currentTimeMillis();
        val MAX: Int = 1000000
        val random: Random = Random()
        val arr: Array<Int> = Array<Int>(n) {i: Int ->
            random.nextInt(MAX + MAX) - MAX
        }
        val count = count(arr)
        val end: Long = System.currentTimeMillis()
        return (end - start) / 1000.0
    }

    fun count(array: Array<Int>): Int {
        val n = array.size - 1
        var count: Int = 0
        for (i: Int in 0..n) {
            for (j: Int in i+1..n) {
                for (k: Int in j+1..n) {
                    if (array[i] + array[j] + array[k] == 0) count++;
                }
            }
        }
        return count
    }
}

fun main(args: Array<String>) {
    val upperBounds: Int = 16000
    var i = 250
    while (i <= upperBounds) {
        var time = DoublingTest().timeTrial(i)
        println("N: $i TIME: $time")
        i *= 2
    }
}